src/go/pgo/distsys/initialization.go

```diff
@@ -0,0 +1,147 @@
+package distsys
+
+import (
+	"net"
+	"net/rpc"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	RPC_ID = "PlusCalProcess"
+)
+
+// This implements the process initialization protocol of PlusCal processes.
+// When a PlusCal specification is compiled to TLA+, the TLC model checker
+// starts every "process" defined in PlusCal at the same time.
+//
+// Process initialization as implemented here makes sure that a process will
+// only start its algorithm execution once all other processes are online.
+// This is particularly useful for specifications that use (or require) fair
+// scheduling.
+
+type ProcessInitialization struct {
+	Peers          []string               // the list of peers (PlusCal processes)
+	Self           string                 // the identification of the current process (IP:port)
+	Coordinator    string                 // which host is the coordinator (IP:port)
+	Network        map[string]*rpc.Client // maps connections from peers
+	readyChan      chan bool              // underlying channel used to coordinate start
+	processesReady int32                  // how many processes are ready currently (used only by the coordinator)
+}
+
+func NewProcessInitialization(peers []string, self, coordinator string) *ProcessInitialization {
+	pi := ProcessInitialization{
+		Peers:       peers,
+		Self:        self,
+		Coordinator: coordinator,
+		Network:     map[string]*rpc.Client{},
+	}
+
+	// if node is the coordinator, create a buffered channel; otherwise, a
+	// synchronous channel is sufficient
+	if self == coordinator {
+		pi.readyChan = make(chan bool, 1)
+	} else {
+		pi.readyChan = make(chan bool)
+	}
+
+	return &pi
+}
+
+func (init *ProcessInitialization) isCoordinator() bool {
+	return init.Self == init.Coordinator
+}
+
+func (self *ProcessInitialization) Init() (*rpc.Server, net.Listener, error) {
```

Comment:
The idea here is to reuse the listener and the server for both the process initialization and the state server. See https://github.com/UBC-NSS/pgo/pull/36/commits/017a85797c0f1d9b1cb6f1fc2099ef891b297bfb for the changes made. If we don't do this, we'll have to allocate a listener, perform the initialization, close the listener, allocate another listener on the same interface, and then resume normal operation.

Reply:
I understand the goal of not binding twice. But we don't need to return the listener from `NewProcessesInitialization` for that -- `NewStateServer` could just **not** call `net.Listen` instead, since it already called `ProcessInitialization.Init()` and that function does the necessary binding.

That said, I see the point of returning the `rpc.Server` instance though, since you want to `RegisterName` on that.

At the end of the day, the state server is still `Dial`ing the other peers when in theory it could reuse the connections on `ProcessInitialization.Network`, and I think that is the more pressing "issue to fix.

I think we need another shared component here that manages RPC server/connections, and that component could be reused across `ProcessInitialization` and `StateServer`. I'll try to come up with an implementation of that.

Reply:
I updated the code with the ideas I mentioned above. Now the state server and process initialization protocol should be reusing their RPC conections, and we don't need to return `rpc.Server` here. Let me know what you think.